### PR TITLE
util: Log early messages

### LINKED ISSUE #16098: LogPrintf() fails due to premature call before InitLogging()
These pieces of code:
https://github.com/bitcoin/bitcoin/blob/8ab4f282c06d67074b872dbda0be37636fdd5186/src/util/system.cpp#L302
and
https://github.com/bitcoin/bitcoin/blob/8ab4f282c06d67074b872dbda0be37636fdd5186/src/util/system.cpp#L864
do not work as expected because they are called before the `InitLogging()` call.

### PR DESCRIPTION
**Pull request by maflcko**
Early log messages are dropped on the floor and they'd never make it to the console or debug log. This can be tested by running the test included in this pull request without re-compiling the `bitcoind`.

Fix that by buffering early messages and flushing them as soon as all logging options have been initialized and logging has been started.

This pull request is identical to  "Log early messages with -printtoconsole" (#13088)  by **ajtowns**, with the following changes:
* Rebased
* Added docstrings for `m_buffering` and `StartLogging`
* Switch `CCriticalSection` (aka `RecursiveMutex`) to just `Mutex` in the last commit
* Added tests

Fixes #16098
Fixes #13157
Closes #13088

### CODE REVIEW

File: src/init.cpp [hunk 1]
```diff
@@ -1249,10 +1249,10 @@ bool AppInitMain(InitInterfaces& interfaces)
             // and because this needs to happen before any other debug.log printing
             LogInstance().ShrinkDebugFile();
         }
-        if (!LogInstance().OpenDebugLog()) {
+    }
+    if (!LogInstance().StartLogging()) {
             return InitError(strprintf("Could not open debug log file %s",
```
Comment (promag) [inline src/init.cpp]: https://github.com/bitcoin/bitcoin/pull/16112/commits/2b0f4e1726379150fd3fc1e3f7bc127303ad850b

nit, could fix indentation.

**Review by promag — COMMENTED** [review]
Tested ACK https://github.com/bitcoin/bitcoin/pull/16112/commits/fa9bc80718b21719452144d63e7c3f7258b06494, ~~I don't see a problem regarding the new circular dependency.~~

File: src/init.cpp (hunk 1 above)
Comment (MarcoFalke) [reply src/init.cpp]: Done

**Comment by MarcoFalke** [comment]
> I don't see a problem regarding the new circular dependency.

It would lead to issues in debug builds when a potential deadlock was detected, so I've reverted that for now.

File: src/init.cpp [hunk 2]
```diff
@@ -1249,10 +1243,10 @@ bool AppInitMain(InitInterfaces& interfaces)
             // and because this needs to happen before any other debug.log printing
             LogInstance().ShrinkDebugFile();
         }
-        if (!LogInstance().OpenDebugLog()) {
-            return InitError(strprintf("Could not open debug log file %s",
+    }
```
Comment (hebasto) [inline src/init.cpp]: Two `if`s above could be combined now.

File: src/logging.cpp [hunk 3]
```diff
@@ -214,32 +229,31 @@ void BCLog::Logger::LogPrintStr(const std::string &str)
 
     m_started_new_line = !str.empty() && str[str.size()-1] == '\n';
```
Comment (hebasto) [inline src/logging.cpp]: May I suggest to replace `str[str.size()-1]` with `str.back()`?

File: src/test/setup_common.cpp [hunk 4]
```diff
@@ -43,10 +43,11 @@ BasicTestingSetup::BasicTestingSetup(const std::string& chainName)
...
     fCheckBlockIndex = true;
     SelectParams(chainName);
-    static bool noui_connected = false;
-    if (!noui_connected) {
+    static bool g_have_singletons_initialized = false;
+    if (!g_have_singletons_initialized) {
+        LogInstance().StartLogging();
         noui_connect();
-        noui_connected = true;
+        g_have_singletons_initialized = true;
```
Comment (hebasto) [inline src/test/setup_common.cpp]: If we mean **single**ton, could it be in singular?

File: src/init.cpp [hunk 5]
```diff
@@ -1249,10 +1243,10 @@ bool AppInitMain(InitInterfaces& interfaces)
...
+    if (!LogInstance().StartLogging()) {
+        return InitError(strprintf("Could not open debug log file %s",
```
Comment (hebasto) [inline src/init.cpp]: Could this message be translated?
Ref: [Translation Strings Policy](https://github.com/bitcoin/bitcoin/blob/master/doc/translation_strings_policy.md#gui-strings)

**Review by hebasto — COMMENTED** [review]
Concept ACK.

**Comment by sipa** [comment]
Concept ACK

**Comment by practicalswift** [comment]
Is there any reason to buffer the writing to `stdout` (`m_print_to_console`)?

**Comment by ajtowns** [comment]
> Is there any reason to buffer the writing to `stdout` (`m_print_to_console`)?

If you have `bitcoind -noconnect=0` but have `printtoconsole=0` in bitcoind.conf, then the "confusing double-negative" error will be queued when parsing the command line, but `m_print_to_console` can't be set to false until later when the config file is actually read.

**Comment by ajtowns** [comment]
I've done #16127 to see what it would take to keep the thread safety annotations without the dependency loop (and hopefully also without the bugs), and what it'd look like to do the same thing elsewhere. Seems better to do that later (or not at all) than to add it to this PR to me...

**Comment by ajtowns** [comment]
utACK faa2a47cd7bcdbd187035c76f8dbd0442f6818dc

**Review by kristapsk — APPROVED** [review]
ACK faa2a47cd7bcdbd187035c76f8dbd0442f6818dc (ran added functional test before / after recompiling, didn't do additional testing)

**Comment by practicalswift** [comment]
@MarcoFalke, why buffer also the writing to `stdout`? Don't we want to print to `stdout` as soon as possible to be on the safe side? I think it makes sense to keep the "print to console" code path as simple/dumb as possible.

> Is there any reason to buffer the writing to stdout (m_print_to_console)?

**Comment by MarcoFalke** [comment]
@practicalswift See https://github.com/bitcoin/bitcoin/pull/16112#issuecomment-497543744

File: src/init.cpp (hunk 2 above)
Comment (MarcoFalke) [reply src/init.cpp]: Leaving it to minimize the diff

File: src/logging.cpp (hunk 3 above)
Comment (MarcoFalke) [reply src/logging.cpp]: Leaving it to minimize the diff

File: src/init.cpp (hunk 5 above)
Comment (MarcoFalke) [reply src/init.cpp]: Leaving it to minimize the diff

File: src/init.cpp (hunk 1 above)
Comment (MarcoFalke) [reply src/init.cpp]: Leaving it to minimize the diff

**Comment by MarcoFalke** [comment]
Unless there are objections, this will be merged tomorrow

**Comment by hebasto** [comment]
> Unless there are objections, this will be merged tomorrow

Going to test in a few hours.

**Comment by hebasto** [comment]
ACK faa2a47cd7bcdbd187035c76f8dbd0442f6818dc

Tested on Linux Mint 19.1:
- the patched `test/functional/feature_config_args.py` fails on master (22b6c4ed7562a23e4363e8f0fd889b92c7653d5f)
- #16098 is fixed:
```
$ cat ~/.bitcoin/bitcoin.conf
trash=1

$ src/qt/bitcoin-qt -noonion=0

$ head ~/.bitcoin/debug.log 





2019-06-18T14:40:23Z Warning: parsed potentially confusing double-negative -onion=0
2019-06-18T14:40:23Z Ignoring unknown configuration value trash
2019-06-18T14:40:23Z Bitcoin Core version v0.18.99.0-10ae3a70e (release build)
2019-06-18T14:40:23Z GUI: QObject::connect: invalid null parameter
2019-06-18T14:40:23Z Assuming ancestors of block 0000000000000000000f1c54590ee18d15ec70e68c8cd4cfbadb1b4f11697eee have valid signatures.
```

Nit: could the test for
https://github.com/bitcoin/bitcoin/blob/8ab4f282c06d67074b872dbda0be37636fdd5186/src/util/system.cpp#L864
be added?

**Pull request merged**